cedar-lean/Cedar/Thm/Data/Map.lean

```diff
@@ -164,31 +372,303 @@ theorem mapOnValues_contains {α β γ} [LT α] [DecidableLT α] [DecidableEq α
+
+theorem mapMOnValues_cons {α : Type 0} [LT α] [DecidableLT α] {f : β → Option γ} {m : Map α β} {k : α} {v : β} {tl : List (α × β)}:
+  m.kvs = (k, v) :: tl →
+  (m.mapMOnValues f = do
+    let v' ← f v
+    let tl' ← (mk tl).mapMOnValues f
+    return mk ((k, v') :: tl'.kvs))
+:= by
+  intro h₁
+  cases h₂ : f v <;> simp only [Option.pure_def, Option.bind_eq_bind, Option.bind_none_fun, Option.bind_some_fun]
+  case none => unfold mapMOnValues ; simp [h₁, h₂]
+  case some v' =>
+    cases h₃ : (mk tl).mapMOnValues f <;> simp only [Option.none_bind, Option.some_bind]
+    <;> unfold mapMOnValues at *
+    case none =>
+      simp only [h₁, Option.pure_def, Option.bind_eq_bind, List.mapM_cons, Option.bind_eq_none,
+        Option.bind_eq_some, Option.some.injEq, forall_exists_index, and_imp,
+        forall_apply_eq_imp_iff₂]
+      intro kvs' v'' h₄ tl' h₅ h₆
+      simp only [h₂, Option.some.injEq] at h₄
+      subst v'' kvs'
+      cases (tl.mapM λ x => match x with | (k, v) => do let v' ← f v ; pure (k, v'))
+      <;> simp only [Option.pure_def, Option.bind_eq_bind, Option.bind_eq_none] at h₃
+      <;> exact h₃ tl' h₅
+    case some mtl' =>
+      simp only [h₁, Option.pure_def, Option.bind_eq_bind, Option.bind_eq_some, Option.some.injEq,
+        List.mapM_cons, mk.injEq, exists_eq_right, List.cons.injEq, exists_eq_right_right,
+        Prod.mk.injEq, true_and] at *
+      apply And.intro h₂
+      replace ⟨tl', h₃, h₄⟩ := h₃
+      subst mtl'
+      simp [h₃]
+
+theorem mapMOnValues_some_implies_forall₂ [LT α] [DecidableLT α] {f : β → Option γ} {m₁ : Map α β} {m₂ : Map α γ} :
+  m₁.mapMOnValues f = some m₂ →
+  List.Forall₂ (λ kv₁ kv₂ => kv₁.fst = kv₂.fst ∧ f kv₁.snd = some kv₂.snd) m₁.kvs m₂.kvs
+:= by
+  unfold mapMOnValues kvs
+  intro h₁
+  simp only [Option.pure_def, Option.bind_eq_bind, Option.bind_eq_some, Option.some.injEq] at h₁
+  replace ⟨x, h₁, h₂⟩ := h₁
+  subst h₂
+  replace h₁ := List.mapM_some_iff_forall₂.mp h₁
+  simp only
+  apply List.Forall₂.imp _ h₁
+  intro (k, v) (k', v') h₂
+  simp only [Option.bind_eq_some, Option.some.injEq, Prod.mk.injEq, exists_eq_right_right] at h₂
+  replace ⟨h₂, h₂'⟩ := h₂
+  subst k'
+  simp only [true_and]
+  exact h₂
+
+theorem mapMOnValues_some_implies_all_some {α : Type 0} [LT α] [DecidableLT α] {f : β → Option γ} {m₁ : Map α β} {m₂ : Map α γ} :
+  m₁.mapMOnValues f = some m₂ →
+  ∀ kv ∈ m₁.kvs, ∃ v, (kv.fst, v) ∈ m₂.kvs ∧ f kv.snd = some v
+:= by
+  intro h₁
+  replace h₁ := List.forall₂_implies_all_left (mapMOnValues_some_implies_forall₂ h₁)
+  intro (k, v) h₂
+  replace ⟨(k', v'), h₁, h₃, h₄⟩ := h₁ (k, v) h₂
+  simp only at *
+  subst k'
+  exists v'
+
+/--
+  alternate proof of `mapMOnValues_some_implies_all_some`, which instead of
+  relying on `mapMOnValues_some_implies_forall₂`, relies on
+  `List.mapM_some_implies_all_some`.  Which do we prefer?
+-/
+theorem mapMOnValues_some_implies_all_some_alt_proof [LT α] [DecidableLT α] {f : β → Option γ} {m₁ : Map α β} {m₂ : Map α γ} :
```

Comment:
Thread for discussion of the two proofs of `mapMOnValues_some_implies_all_some`

Reply:
Same comment as above :). No strong feelings but my preference is to have fewer specialized lemmas that aren't used anywhere else.  (Obviously, there is a limit to this rule-of-thumb, and I do have super long proofs that are broken into many single-use lemmas---though in that case, I make those `private`.)

Reply:
Wait is there a `private` keyword in Lean?

In this case, unlike in the above case, neither of the lemmas these proofs rely on exist solely for this proof's benefit.  Schematically, we have
```
A ---- B
|      |
C ---- D
```
This is A, we need proofs of A,B,C,D, and we can choose whether to prove A "through" B or "through" C.  

(In the figure, the top row is Map lemmas, the bottom row is List lemmas, the left side is mapM[OnValues] lemmas, the right side is `forall2` lemmas.)

Reply:
Yeah, the `private` keyword is great:  it limits the visibility of the theorem / definition to the defining module.  

Wish it worked for macros!